cvejob/version_utils.py

```diff
@@ -0,0 +1,336 @@
+            return '{hi}'.format(hi=self.boundary_1)
+        elif self.boundary_1 <= self.boundary_2:
+            return '{hi},{lo}'.format(hi=self.boundary_2, lo=self.boundary_1.version)
+        else:
+            return '{hi},{lo}'.format(hi=self.boundary_1, lo=self.boundary_2.version)
+
+    def __repr__(self):
+        return 'VersionRange({r})'.format(r=str(self))
+
+
+def classify_versions(upstream_versions, version_ranges):
+    """Classify versions as affected and unaffected.
+
+    :param upstream_versions: list[str], list of upstream versions
+    :param version_ranges: list[[VersionRange]], list of CVE version ranges
+
+    :return: list[ClassifiedVersion] sorted list of classified upstream versions
+    """
+    sorted_upstream_versions = [ComparableVersion(x) for x in upstream_versions]
+    sorted_upstream_versions.sort()
+
+    marked_versions = [ClassifiedVersion(x.version, False) for x in sorted_upstream_versions]
+
+    for version_tuple in marked_versions:
+        benev_upstream_version = BenevolentVersion(version_tuple.version)
+
+        for version_range in version_ranges:
+            if version_range.contains(benev_upstream_version.exact, cmp_class=BenevolentVersion):
+                version_tuple.is_affected = True
+                break
+
+    return marked_versions
+
+
+def get_affected_ranges(classified_versions):
+    """Get version ranges for affected versions.
+
+    :param classified_versions: [ClassifiedVersion], sorted list of classified versions
+    :return: list of VersionRange objects
+    """
+    unaffected_found = False
+    affected_list = []
+
+    ranges = []
+
+    def append_range(affected_versions):
+        if affected_versions:
+            ranges.append(
+                VersionRange.from_list(
+                    affected_versions, is_left_closed=unaffected_found, is_right_closed=True
+                )
+            )
+
+    for version_tuple in classified_versions:
+        if not version_tuple.is_affected:
+            append_range(affected_list)
+            unaffected_found = True
+            affected_list.clear()
+        else:
+            affected_list.append(version_tuple.version)
+    append_range(affected_list)
+
+    return ranges
+
+
+def get_safe_ranges(classified_versions):
+    """Get version ranges for unaffected versions.
+
+    :param classified_versions: [ClassifiedVersion], sorted list of classified versions
+    :return: list of VersionRange objects
+    """
+    affected_found = False
+    safe_list = []
+
+    ranges = []
+
+    def append_range(safe_versions):
+        if safe_versions:
+            ranges.append(
+                VersionRange.from_list(
+                    safe_versions, is_left_closed=True, is_right_closed=affected_found
+                )
+            )
+
+    for version_tuple in reversed(classified_versions):
+        if version_tuple.is_affected:
+            append_range(safe_list)
+            affected_found = True
```

Comment:
If I understand this correctly, once one affected version is found, all of the previous ones are marked as `is_right_closed=True`. Is it intended behaviour? Shouldn't the `affected_found` be falsified again after assignment?

Reply:
Thanks, this probably deserves some comments in the code. This is intended behaviour. We iterate over reversed list of sorted upstream versions. We don't want the interval to be right-closed at the current latest (safe) version. Only when we find the first affected sub-interval, only then we close the right side for safe versions. 

Basically, we want something like this (4.1.3 is the latest version, and it is unaffected):
```
fixedin:
    - ">=4.1.3"
    - "<=3.5.0,3.4.2"
    ...
```